// mochapack/lib/webpack/loader/entryLoader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var loader_utils_1 = __importDefault(require("loader-utils"));
var normalize_path_1 = __importDefault(require("normalize-path"));
var createEntry_1 = __importDefault(require("../util/createEntry"));
var EntryConfig = /** @class */ (function () {
    function EntryConfig() {
        this.files = [];
    }
    EntryConfig.prototype.addFile = function (file) {
        var normalizedFile = normalize_path_1["default"](file);
        this.files.push(normalizedFile);
    };
    EntryConfig.prototype.removeFile = function (file) {
        var normalizedFile = normalize_path_1["default"](file);
        this.files = this.files.filter(function (f) { return f !== normalizedFile; });
    };
    EntryConfig.prototype.getFiles = function () {
        return this.files;
    };
    return EntryConfig;
}());
exports.EntryConfig = EntryConfig;
exports.entryLoader = function entryLoader() {
    var _this = this;
    var loaderOptions = loader_utils_1["default"].getOptions(this);
    var config = loaderOptions.entryConfig;
    // Remove all dependencies of the loader result
    this.clearDependencies();
    var dependencies = config
        .getFiles()
        .map(function (file) { return loader_utils_1["default"].stringifyRequest(_this, file); });
    // add all entries as dependencies
    dependencies.forEach(this.addDependency.bind(this));
    // build source code
    var sourceCode = createEntry_1["default"](dependencies);
    this.callback(null, sourceCode, null);
};
exports["default"] = exports.entryLoader;
