// vue-tailwind/dist/l10n/at.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Austria = void 0;
exports.Austria = {
    weekdays: {
        shorthand: ['So', 'Mo', 'Di', 'Mi', 'Do', 'Fr', 'Sa'],
        longhand: [
            'Sonntag',
            'Montag',
            'Dienstag',
            'Mittwoch',
            'Donnerstag',
            'Freitag',
            'Samstag',
        ],
    },
    months: {
        shorthand: [
            'J채n',
            'Feb',
            'M채r',
            'Apr',
            'Mai',
            'Jun',
            'Jul',
            'Aug',
            'Sep',
            'Okt',
            'Nov',
            'Dez',
        ],
        longhand: [
            'J채nner',
            'Februar',
            'M채rz',
            'April',
            'Mai',
            'Juni',
            'Juli',
            'August',
            'September',
            'Oktober',
            'November',
            'Dezember',
        ],
    },
    firstDayOfWeek: 1,
    weekAbbreviation: 'KW',
    rangeSeparator: ' bis ',
};
exports.default = exports.Austria;
