// vue-tailwind/dist/l10n/bs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bosnian = void 0;
exports.Bosnian = {
    firstDayOfWeek: 1,
    weekdays: {
        shorthand: ['Ned', 'Pon', 'Uto', 'Sri', 'Čet', 'Pet', 'Sub'],
        longhand: [
            'Nedjelja',
            'Ponedjeljak',
            'Utorak',
            'Srijeda',
            'Četvrtak',
            'Petak',
            'Subota',
        ],
    },
    months: {
        shorthand: [
            'Jan',
            'Feb',
            'Mar',
            'Apr',
            'Maj',
            'Jun',
            'Jul',
            'Avg',
            'Sep',
            'Okt',
            'Nov',
            'Dec',
        ],
        longhand: [
            'Januar',
            'Februar',
            'Mart',
            'April',
            'Maj',
            'Juni',
            'Juli',
            'Avgust',
            'Septembar',
            'Oktobar',
            'Novembar',
            'Decembar',
        ],
    },
    time24hr: true,
};
exports.default = exports.Bosnian;
