// vue-tailwind/dist/l10n/fa.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Persian = void 0;
exports.Persian = {
    weekdays: {
        shorthand: ['یک', 'دو', 'سه', 'چهار', 'پنج', 'جمعه', 'شنبه'],
        longhand: [
            'یک‌شنبه',
            'دوشنبه',
            'سه‌شنبه',
            'چهارشنبه',
            'پنچ‌شنبه',
            'جمعه',
            'شنبه',
        ],
    },
    months: {
        shorthand: [
            'ژانویه',
            'فوریه',
            'مارس',
            'آوریل',
            'مه',
            'ژوئن',
            'ژوئیه',
            'اوت',
            'سپتامبر',
            'اکتبر',
            'نوامبر',
            'دسامبر',
        ],
        longhand: [
            'ژانویه',
            'فوریه',
            'مارس',
            'آوریل',
            'مه',
            'ژوئن',
            'ژوئیه',
            'اوت',
            'سپتامبر',
            'اکتبر',
            'نوامبر',
            'دسامبر',
        ],
    },
    firstDayOfWeek: 6,
    ordinal: () => '',
};
exports.default = exports.Persian;
