// vue-tailwind/dist/l10n/lv.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Latvian = void 0;
exports.Latvian = {
    firstDayOfWeek: 1,
    weekdays: {
        shorthand: ['Sv', 'Pr', 'Ot', 'Tr', 'Ce', 'Pk', 'Se'],
        longhand: [
            'Svētdiena',
            'Pirmdiena',
            'Otrdiena',
            'Trešdiena',
            'Ceturtdiena',
            'Piektdiena',
            'Sestdiena',
        ],
    },
    months: {
        shorthand: [
            'Jan',
            'Feb',
            'Mar',
            'Apr',
            'Mai',
            'Jūn',
            'Jūl',
            'Aug',
            'Sep',
            'Okt',
            'Nov',
            'Dec',
        ],
        longhand: [
            'Janvāris',
            'Februāris',
            'Marts',
            'Aprīlis',
            'Maijs',
            'Jūnijs',
            'Jūlijs',
            'Augusts',
            'Septembris',
            'Oktobris',
            'Novembris',
            'Decembris',
        ],
    },
    rangeSeparator: ' līdz ',
    time24hr: true,
};
exports.default = exports.Latvian;
