// vue-tailwind/dist/l10n/ms.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Malaysian = void 0;
exports.Malaysian = {
    weekdays: {
        shorthand: ['Min', 'Isn', 'Sel', 'Rab', 'Kha', 'Jum', 'Sab'],
        longhand: [
            'Minggu',
            'Isnin',
            'Selasa',
            'Rabu',
            'Khamis',
            'Jumaat',
            'Sabtu',
        ],
    },
    months: {
        shorthand: [
            'Jan',
            'Feb',
            'Mac',
            'Apr',
            'Mei',
            'Jun',
            'Jul',
            'Ogo',
            'Sep',
            'Okt',
            'Nov',
            'Dis',
        ],
        longhand: [
            'Januari',
            'Februari',
            'Mac',
            'April',
            'Mei',
            'Jun',
            'Julai',
            'Ogos',
            'September',
            'Oktober',
            'November',
            'Disember',
        ],
    },
    firstDayOfWeek: 1,
    ordinal: () => '',
};
exports.default = exports.Malaysian;
