// vue-tailwind/dist/l10n/sr-cyr.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SerbianCyrillic = void 0;
exports.SerbianCyrillic = {
    weekdays: {
        shorthand: ['Нед', 'Пон', 'Уто', 'Сре', 'Чет', 'Пет', 'Суб'],
        longhand: [
            'Недеља',
            'Понедељак',
            'Уторак',
            'Среда',
            'Четвртак',
            'Петак',
            'Субота',
        ],
    },
    months: {
        shorthand: [
            'Јан',
            'Феб',
            'Мар',
            'Апр',
            'Мај',
            'Јун',
            'Јул',
            'Авг',
            'Сеп',
            'Окт',
            'Нов',
            'Дец',
        ],
        longhand: [
            'Јануар',
            'Фебруар',
            'Март',
            'Април',
            'Мај',
            'Јун',
            'Јул',
            'Август',
            'Септембар',
            'Октобар',
            'Новембар',
            'Децембар',
        ],
    },
    firstDayOfWeek: 1,
    weekAbbreviation: 'Нед.',
    rangeSeparator: ' до ',
};
exports.default = exports.SerbianCyrillic;
