// vue-tailwind/dist/l10n/uk.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ukrainian = void 0;
exports.Ukrainian = {
    firstDayOfWeek: 1,
    weekdays: {
        shorthand: ['Нд', 'Пн', 'Вт', 'Ср', 'Чт', 'Пт', 'Сб'],
        longhand: [
            'Неділя',
            'Понеділок',
            'Вівторок',
            'Середа',
            'Четвер',
            "П'ятниця",
            'Субота',
        ],
    },
    months: {
        shorthand: [
            'Січ',
            'Лют',
            'Бер',
            'Кві',
            'Тра',
            'Чер',
            'Лип',
            'Сер',
            'Вер',
            'Жов',
            'Лис',
            'Гру',
        ],
        longhand: [
            'Січень',
            'Лютий',
            'Березень',
            'Квітень',
            'Травень',
            'Червень',
            'Липень',
            'Серпень',
            'Вересень',
            'Жовтень',
            'Листопад',
            'Грудень',
        ],
    },
    time24hr: true,
};
exports.default = exports.Ukrainian;
