// typechain/dist/cli/cli.js
#!/usr/bin/env node
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const prettier = __importStar(require("prettier"));
const runTypeChain_1 = require("../typechain/runTypeChain");
const glob_1 = require("../utils/glob");
const logger_1 = require("../utils/logger");
const parseArgs_1 = require("./parseArgs");
async function main() {
    ;
    global.IS_CLI = true;
    const cliConfig = parseArgs_1.parseArgs();
    const cwd = process.cwd();
    const config = {
        cwd,
        target: cliConfig.target,
        outDir: cliConfig.outDir,
        allFiles: glob_1.glob(cwd, cliConfig.files),
        filesToProcess: glob_1.glob(cwd, cliConfig.files),
        prettier,
        flags: {
            ...cliConfig.flags,
            environment: undefined,
        },
    };
    const result = await runTypeChain_1.runTypeChain(config);
    console.log(`Successfully generated ${result.filesGenerated} typings!`);
}
main().catch((e) => {
    logger_1.logger.error('Error occured: ', e.message);
    const stackTracesEnabled = process.argv.includes('--show-stack-traces');
    if (stackTracesEnabled) {
        logger_1.logger.error('Stack trace: ', e.stack);
    }
    else {
        logger_1.logger.error('Run with --show-stack-traces to see the full stacktrace');
    }
    process.exit(1);
});
