// typechain/dist/cli/parseArgs.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseArgs = void 0;
const command_line_args_1 = __importDefault(require("command-line-args"));
const DEFAULT_GLOB_PATTERN = '**/*.abi';
function parseArgs() {
    const optionDefinitions = [
        { name: 'glob', type: String, defaultOption: true, multiple: true },
        { name: 'target', type: String },
        { name: 'out-dir', type: String },
        { name: 'show-stack-traces', type: Boolean },
        { name: 'always-generate-overloads', type: Boolean },
    ];
    const rawOptions = command_line_args_1.default(optionDefinitions);
    return {
        files: rawOptions.glob || [DEFAULT_GLOB_PATTERN],
        outDir: rawOptions['out-dir'],
        target: rawOptions.target,
        flags: {
            alwaysGenerateOverloads: rawOptions['always-generate-overloads'] || false,
        },
    };
}
exports.parseArgs = parseArgs;
