// mochapack/lib/cli/requireWebpackConfig.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var path_1 = __importDefault(require("path"));
var fs_1 = __importDefault(require("fs"));
var interpret_1 = __importDefault(require("interpret"));
function sortExtensions(ext1, ext2) {
    if (ext1 === '.js') {
        return -1;
    }
    if (ext2 === '.js') {
        return 1;
    }
    return ext1.length - ext2.length;
}
var extensions = Object.keys(interpret_1["default"].extensions).sort(sortExtensions);
function fileExists(filePath) {
    try {
        return fs_1["default"].existsSync(filePath);
    }
    catch (e) {
        return false;
    }
}
function findConfigFile(dirPath, baseName) {
    for (var i = 0; i < extensions.length; i += 1) {
        var filePath = path_1["default"].resolve(dirPath, "" + baseName + extensions[i]);
        if (fileExists(filePath)) {
            return filePath;
        }
    }
    return null;
}
function getConfigExtension(configPath) {
    for (var i = extensions.length - 1; i >= 0; i -= 1) {
        var extension = extensions[i];
        if (configPath.indexOf(extension, configPath.length - extension.length) > -1) {
            return extension;
        }
    }
    return path_1["default"].extname(configPath);
}
function registerCompiler(moduleDescriptor) {
    if (!moduleDescriptor) {
        return;
    }
    if (typeof moduleDescriptor === 'string') {
        require(moduleDescriptor); // eslint-disable-line global-require, import/no-dynamic-require
    }
    else if (!Array.isArray(moduleDescriptor)) {
        var module_1 = require(moduleDescriptor.module); // eslint-disable-line global-require, import/no-dynamic-require
        moduleDescriptor.register(module_1);
    }
    else {
        for (var i = 0; i < moduleDescriptor.length; i += 1) {
            try {
                registerCompiler(moduleDescriptor[i]);
                break;
            }
            catch (e) {
                // do nothing
            }
        }
    }
}
function requireWebpackConfig(webpackConfig, required, env, mode) {
    return __awaiter(this, void 0, void 0, function () {
        var configPath, configExtension, configFound, config, configDirPath, configBaseName, configPathPrecompiled, configExtensionPrecompiled;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    configPath = path_1["default"].resolve(webpackConfig);
                    configExtension = getConfigExtension(configPath);
                    configFound = false;
                    config = {};
                    if (fileExists(configPath)) {
                        // config exists, register compiler for non-js extensions
                        registerCompiler(interpret_1["default"].extensions[configExtension]);
                        // require config
                        config = require(configPath); // eslint-disable-line global-require, import/no-dynamic-require
                        configFound = true;
                    }
                    else if (configExtension === '.js') {
                        configDirPath = path_1["default"].dirname(configPath);
                        configBaseName = path_1["default"].basename(configPath, configExtension);
                        configPathPrecompiled = findConfigFile(configDirPath, configBaseName);
                        if (configPathPrecompiled != null) {
                            configExtensionPrecompiled = getConfigExtension(configPathPrecompiled);
                            // register compiler & require config
                            registerCompiler(interpret_1["default"].extensions[configExtensionPrecompiled]);
                            config = require(configPathPrecompiled); // eslint-disable-line global-require, import/no-dynamic-require
                            configFound = true;
                        }
                    }
                    if (!configFound) {
                        if (required) {
                            throw new Error("Webpack config could not be found: " + webpackConfig);
                        }
                        else if (mode != null) {
                            ;
                            config.mode = mode;
                        }
                        return [2 /*return*/, config];
                    }
                    // @ts-ignore
                    config = config["default"] || config;
                    if (!(typeof config === 'function')) return [3 /*break*/, 2];
                    return [4 /*yield*/, Promise.resolve(config(env))];
                case 1:
                    config = _a.sent();
                    _a.label = 2;
                case 2:
                    if (mode != null) {
                        config.mode = mode;
                    }
                    if (Array.isArray(config)) {
                        throw new Error('Passing multiple configs as an Array is not supported. Please provide a single config instead.');
                    }
                    return [2 /*return*/, config];
            }
        });
    });
}
exports["default"] = requireWebpackConfig;
